Editor/EditorCore/CutTool.cs

```diff
@@ -0,0 +1,1421 @@
+
+                SceneView.RepaintAll();
+            }
+            //If the user is adding the current position to the cut
+            else if (evtType == EventType.MouseDown
+                     && HandleUtility.nearestControl == m_ControlId)
+            {
+                int polyCount = m_CutPath.Count;
+                if (!m_CurrentPosition.Equals(Vector3.positiveInfinity)
+                    && (polyCount == 0 || m_SelectedIndex != polyCount - 1))
+                {
+                    AddCurrentPositionToPath();
+                    evt.Use();
+                }
+            }
+        }
+
+        internal void AddCurrentPositionToPath()
+        {
+            UndoUtility.RecordObject(this, "Add Vertex On Path");
+
+            if(m_TargetFace == null)
+                m_TargetFace = m_CurrentFace;
+
+            m_CutPath.Add(new CutVertexData(m_CurrentPosition, m_CurrentPositionNormal, m_CurrentVertexTypes));
+            UpdateMeshConnections();
+
+            m_PlacingPoint = true;
+            m_SelectedIndex = m_CutPath.Count - 1;
+
+            RebuildCutShape();
+        }
+
+        /// <summary>
+        /// Compute the position designated by the user in the current mesh/face taking into account snapping
+        /// </summary>
+        /// <returns>true is a valid position is computed in the mesh</returns>
+        bool UpdateHitPosition()
+        {
+            Event evt = Event.current;
+
+            Ray ray = UHandleUtility.GUIPointToWorldRay(evt.mousePosition);
+            RaycastHit pbHit;
+
+            m_CurrentFace = null;
+
+            if (UnityEngine.ProBuilder.HandleUtility.FaceRaycast(ray, m_Mesh, out pbHit))
+            {
+                UpdateCurrentPosition(m_Mesh.faces[pbHit.face], pbHit.point ,pbHit.normal);
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Add the position in the face to the cut taking into account snapping
+        /// </summary>
+        internal void UpdateCurrentPosition(Face face, Vector3 position, Vector3 normal)
+        {
+            m_CurrentPosition = position;
+            m_CurrentPositionNormal = normal;
+            m_CurrentFace = face;
+            m_CurrentVertexTypes = VertexTypes.None;
+
+            CheckPointInCutPath();
+
+            if (m_CurrentVertexTypes == VertexTypes.None && !m_ModifyingPoint)
+                CheckPointInMesh();
+        }
+
+        /// <summary>
+        /// Updates the connections between the cut path and the mesh vertices
+        /// </summary>
+        internal void UpdateMeshConnections()
+        {
+            m_MeshConnections.Clear();
+            if(m_CutPath.Count < 2)
+                return;
+
+            List<Vector3> existingVerticesInCut =
+                m_CutPath.Where(v => ( v.types | VertexTypes.ExistingVertex ) != 0)
+                         .Select(v => v.position).ToList();
+
+            Vertex[] vertices = m_Mesh.GetVertices();
```

Comment:
if you're not adding or removing vertices you can save an alloc here by using `m_Mesh.positionsInternal`